zephyr: page frame management, demand paging, x86 implementation

tyrantlizard77 opened the pull request:
This PR:

- Implements physical page frame management for the core kernel. All pages in main system ram now have an associated `struct z_page_frame` for their metadata. The kernel marks only the page frames containing the boot zephyr image as being used, with remaining pages placed on a free page frames list.
- Support for anonymous memory mappings with the `k_mem_map()` API. These will be mapped into the virtual address space using free physical page frames. The newlib `malloc()` heap has been updated to use a chunk of anonymous memory.
- A policy change has been implemented for boot memory mappings. Only the zephyr image (all text + rodata + data + bss + noinit) is now mapped at boot. Remaining RAM pages are not mapped and placed on the free pages list.
- A policy change has been implemented for mapping memory: the virtual address space must be pre-allocated at boot. This makes `arch_mem_map()` an operation which cannot fail due to insufficient memory for paging structures. The only runtime allocations made with respect to paging are the instantiation of new memory domains on arches which are not configured to use a single page table.
- New architecture interfaces have been specified to support demand paging. APIs for querying page table information and updating page tables for page-ins/page-outs are defined.
- The core kernel now has a hook to handle page faults, which will be invoked by the architecture appropriately.
- Interfaces for backing store implementations are now defined. Real-world use-cases will tend to have custom implementations that are closely tied to target hardware, but a sample implementation which creates a RAM-based backing store is implemented for testing purposes.
- Interfaces for eviction algorithms are now defined. Zephyr provides a simple Not Recently Used eviction algorithm.
- APIs to give applications control of demand paging policy are defined. Pages may be pinned in memory, un-pinned, and paged-in/out by virtual address.
- Implementation on x86 which supports all the architecture-specific bits mentioned above.
- Basic functionality and API tests

To be implemented in further work in the 2.6 cycle:

- this implementation currently pins the whole kernel image, only implementing demand paging for anonymous pages. Subsequent work will pin only those parts of the kernel which are needed for pagefault handling and support pinning application pages at build time.
- Implement performance metrics (pagefault counts, latencies, etc)

Commit:
arch: add CONFIG_DEMAND_PAGING

Indicates at the kernel level that demand paging is active.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add demand paging arch interfaces

Architecture layer hooks for demand paging. See
doxygen for these API definitions for more details.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add CONFIG_ARCH_HAS_RESERVED_PAGE_FRAMES

We will need this to run on x86 with PC-like hardware.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
demand_paging: add infra for demand paging modules

Backing stores and eviction algorithms will be included here.
Exactly one must be chosen, with a default option to leave
the implementation to the application.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add app-facing demand paging APIs

Routines to evict memory, page-in memory, and set pinned state
at runtime.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: implement demand paging APIs

All arch_ APIs and macros are implemented, and the page fault
handling code will call into the core kernel.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
qemu_x86_tiny: enable demand paging

This target is specifically for simulating x86 micro-
controllers with limited memory.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add page frame management

Initialize the page frame ontology at boot and update it
when we do memory mappings.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
demand_paging: add NRU algorithm

Simple textbook Not Recently Used eviction algorithm.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
demand_paging: add RAM-based demo backing store

Will be used by QEMU targets for testing purposes.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: reserve the first megabyte

A more comprehensive solution would use E820 enumeration, but we
are unlikely to ever care that much, as we intend to use demand
paging on microcontrollers and not PC-like hardware. This is
really to just prevent QEMU from crashing.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: only map the kernel image

The policy is changed and we no longer map all page frames.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
linker-defs: add syms for kernel image bounds

These are needed on MMU systems and define where the kernel
image resides in virtual memory at boot so that it may be
memory-mapped.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: linker: define z_mapped_* symbols

We will use this to map the kernel instead of all RAM.
The end of the kernel is always page-aligned, regardless
of CONFIG_SRAM_REGION_PERMISSIONS as it must be mapped.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: mem_map: pin test pages

These get mapped to multiple virtual addresses and must be
pinned.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add demand paging internal interfaces

APIs used by backing store and eviction algorithms.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add demand paging implementation

Implement runtime APIs for pinning, paging in, and evicting
memory, as well as the page fault hook called from architecture
code.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
qemu_x86_tiny: don't use first megabyte at all

Just tell the kernel that RAM starts 1MB in, period.
Better simulation of a low-memory microcontroller as
we're not managing a very large number of page frames
we'll never use.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add k_mem_map() interface

Allows applications to increase the data space available to Zephyr
via anonymous memory mappings. Loosely based on mmap().

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: x86: pagetables: pass if userspace disabled

_app_smem doesn't exist in this case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: tests: pagetables: fix assumptions

All RAM may not be mapped. Check the mapping for the main kernel
image and the locore if it exists.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arch: remove KERNEL_RAM_SIZE

We don't map all RAM at boot any more, just the kernel image.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: pre-allocate address space

We no longer use a page pool to draw memory pages when doing
memory map operations. We now preallocate the entire virtual
address space so no allocations are ever necessary when mapping
memory.

We still need memory to clone page tables, but this is now
expressed by a new Kconfig X86_MAX_ADDITIONAL_MEM_DOMAINS
which has much clearer semantics than specifying the number
of pages in the pool.

The default address space size is now 8MB, but this can be
tuned by the application.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arch: add KERNEL_VM_OFFSET

This is only needed if the base address of SRAM doesn't
have the same alignment as the base address of the virtual
address space.

Fix the calculations on X86 where this is the case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
newlib: memory-map the heap, cleanups

We now draw heap memory from an anonymous memory mapping
instead of a hard-coded region past the kernel image,
which is no longer mapped by default.

Some readability cleanups were made to a particuarly
horrible set of nested ifdefs. A few types were adjusted.

sbrk()'s count argument is an intptr_t, not an int.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
mmu: add k_mem_free_get()

Return the amount of physical anonymous memory remaining.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
mmu: arch_mem_map() may no longer fail

Pre-allocation of paging structures is now required, such that
no allocations are ever needed when mapping memory.

Instantiation of new memory domains may still require allocations
unless a common page table is used.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
mmu: ensure gperf data is mapped

Page tables created at build time may not include the
gperf data at the very end of RAM. Ensure this is mapped
properly at runtime to work around this.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
mmu: pin the whole kernel

This will enable testing of the implementation until the
critical set of pages is identified and known to the
kernel.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: add basic k_mem_map() test

Show we can measure free memory properly and map a page of
anonymous memory, which has been zeroed and is writable.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: add intial demand paging testcase

More to be added, but for now show that we can map more
anonymous memory than we physically have, and that reading/
writing to it works as expected.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
This is now working end-to-end, my focus is now documentation and test cases for existing functionality with the aim of taking this out of draft status later this week.

Commit:
arm64: linker: define z_mapped_* symbols

Add linker symbols corresponding to the start and end of the
mapped Zephyr image. This is not used by the ARM arch yet, but
is required to compile the core kernel MMU code.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
newlib: clamp max heap size on MMU systems

Previously, newlib claimed all free physical memory in the
system.

Now, the kernel manages this, allowing for memory to be
used via k_mem_map() calls.

Establish an upper bound to how much newlib will try to
claim on system startup, instead of trying to take all
of it, allowing other parts of the system to also map
anonymous memory.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
CODEOWNERS: add demand paging subdir

Adding myself as original author.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add z_num_pagefaults_get()

Simple counter of number of successfully handled page faults by
the core kernel.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: demand_paging: add more API tests

Add remaining APIs in mem_manage.h even though not all have
been promoted to public (yet).

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
This is ready for review.

There is one failing test. tests/kernel/context fails in qemu_x86_tiny due to the periodic timer hook in the NRU algorithm. I was able to reproduce the problem in the master branch and I think it's a kernel bug with timeouts. I opened https://github.com/zephyrproject-rtos/zephyr/issues/31333

Commit:
mmu: backing stores reserve page fault room

If we evict enough pages to completely fill the backing store,
through APIs like k_mem_map(), z_page_frame_evict(), or
z_mem_page_out(), this will produce a crash the next time we
try to handle a page fault.

The backing store now always reserves a free storage location
for actual page faults.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: context: disable if DEMAND_PAGING

Until #31333 is resolved, the periodic timer in the eviction
algorithm interacts with this test in such a way that the system
deadlocks.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

dcpleung commented on tests/kernel/mem_protect/demand_paging/src/main.c:
```diff
@@ -74,12 +86,112 @@ void test_touch_anon_pages(void)
+
+	zassert_equal(faults, HALF_PAGES,
+		      "unexpected num pagefaults expected %lu got %d",
+		      faults, HALF_PAGES);
```
I think the arguments are reversed. Should be "expected `HALF_PAGES` got `faults`".

andrewboie commented on include/linker/kobject.ld:
```diff
@@ -48,4 +49,5 @@
 	z_priv_stacks_ram_end = .;
 	} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
 #endif /* CONFIG_GEN_PRIV_STACKS */
+	_kobject_data_end = .;
```
oops, this is actually unused, removing.

nashif commented on include/sys/mem_manage.h:
```diff
@@ -204,6 +204,75 @@ void *k_mem_map(size_t size, uint32_t flags);
+ * @retval -ENOMEM Insufficient space in backing store to satisfy request.
+ *         The region may be partially paged out.
+ */
+int z_mem_page_out(void *addr, size_t size);
```
if those are app-facing APIs, shouldn't this be mem_page_out(..), mem_page_in(..) etc?

Once we start adding documentation for paging in general and start referencing those APIs it will not work. Any functions starting with z_ will not be processed by doxygen.

nashif replied:
You should not be concerned about deprecation policy here, those APIs will go in as experimental/unstable and the deprecation policy only applies when this went through 2 releases and when more architectures are supported, so we should be good to move to k_mem_...

andrewboie replied:
promoted.

Commit:
mmu: promote public APIs

These are application facing and are prefixed with k_.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

jhedberg commented:
It seems like this broke both UP2 and EHL: #31517

commit 3660040 which both I and @tbursztyka bisected this to doesn't seem to be independently revertible, and since this PR has so many commits, it's probably wiser to move forward with a fix rather than just reverting everything.
